File: Pastura/PasturaTests/Engine/ScoringLogic/WordwolfJudgeLogicTests.swift
import Testing

@testable import Pastura

@Suite(.timeLimit(.minutes(1)))
struct WordwolfJudgeLogicTests {
  let logic = WordwolfJudgeLogic()

  @Test func majorityWinsWhenWolfMostVoted() {
    var state = SimulationState()
    state.voteResults = ["Wolf": 3, "Other": 1]
    state.variables["wolf_name"] = "Wolf"
    let collector = EventCollector()
    logic.calculate(state: &state, language: "ja", emitter: collector.emit)
    let summaries = collector.events.compactMap { event -> String? in
      if case .summary(let text) = event { return text }
      return nil
    }
    #expect(summaries.count == 1)
    #expect(summaries[0].contains("多数派の勝ち"))
  }

  @Test func wolfWinsWhenNotDetected() {
    var state = SimulationState()
    state.voteResults = ["Innocent": 3, "Wolf": 1]
    state.variables["wolf_name"] = "Wolf"
    let collector = EventCollector()
    logic.calculate(state: &state, language: "ja", emitter: collector.emit)
    let summaries = collector.events.compactMap { event -> String? in
      if case .summary(let text) = event { return text }
      return nil
    }
    #expect(summaries[0].contains("ウルフの勝ち"))
  }

  /// A vote tie must resolve to a **stable** winner. Before #1057 the winner
  /// came from `Dictionary.max(by:)` with no secondary key, so it flipped with
  /// hash-seed randomization across launches. Canonical tie-break is
  /// (count desc, name desc) → Bob wins the tie; with the wolf being Alice,
  /// the wolf escapes deterministically on every call.
  @Test func tieResolvesToStableWinner() {
    func run() -> String {
      var state = SimulationState()
      state.voteResults = ["Alice": 2, "Bob": 2]
      state.variables["wolf_name"] = "Alice"
      let collector = EventCollector()
      logic.calculate(state: &state, language: "ja", emitter: collector.emit)
      let summaries = collector.events.compactMap { event -> String? in
        if case .summary(let text) = event { return text }
        return nil
      }
      return summaries.first ?? ""
    }

    let first = run()
    // Load-bearing: the deterministic winner is Bob (name desc); Bob != wolf
    // (Alice) → wolf wins. The old `max(by:)` could not pin this — it returned
    // whichever tied key dictionary order surfaced first.
    #expect(first.contains("Bob"))
    #expect(first.contains("ウルフの勝ち"))
    #expect(!first.contains("Alice"))
    // Independently-constructed states yield the identical outcome. Pre-fix
    // these could diverge within a single run (observed Alice, then Bob);
    // the sort makes the winner a pure function of the vote counts + names.
    #expect(run() == first)
    #expect(run() == first)
  }

  @Test func handlesEmptyVoteResults() {
    var state = SimulationState()
    let collector = EventCollector()
    logic.calculate(state: &state, language: "ja", emitter: collector.emit)
    let summaries = collector.events.compactMap { event -> String? in
      if case .summary(let text) = event { return text }
      return nil
    }
    #expect(summaries[0].contains("投票結果がありません"))
  }

  // MARK: - English language path (ADR-010 Step E)

  @Test func majorityWinsInEnglish() {
    var state = SimulationState()
    state.voteResults = ["Wolf": 3, "Other": 1]
    state.variables["wolf_name"] = "Wolf"
    let collector = EventCollector()
    logic.calculate(state: &state, language: "en", emitter: collector.emit)
    let summaries = collector.events.compactMap { event -> String? in
      if case .summary(let text) = event { return text }
      return nil
    }
    #expect(summaries.count == 1)
    #expect(summaries[0].contains("Majority wins"))
    #expect(!summaries[0].contains("多数派の勝ち"))
  }

  @Test func wolfWinsInEnglish() {
    var state = SimulationState()
    state.voteResults = ["Innocent": 3, "Wolf": 1]
    state.variables["wolf_name"] = "Wolf"
    let collector = EventCollector()
    logic.calculate(state: &state, language: "en", emitter: collector.emit)
    let summaries = collector.events.compactMap { event -> String? in
      if case .summary(let text) = event { return text }
      return nil
    }
    #expect(summaries.count == 1)
    #expect(summaries[0].contains("The wolf wins"))
    #expect(!summaries[0].contains("ウルフの勝ち"))
  }
}

File: shared/engine/src/commonTest/kotlin/com/pastura/engine/ScoringLogic/WordwolfJudgeLogicTests.kt
package com.pastura.engine

import com.pastura.models.SimulationEvent
import com.pastura.models.SimulationState
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

/**
 * Kotlin sibling of Swift's `WordwolfJudgeLogicTests`.
 *
 * Includes rendered-string-equality assertions (ja + en) that pin the Swift
 * `String(format:)` → Kotlin interpolation parity.
 *
 * Ported for the ADR-023 Stage-3 PR-1 score_calc slice (#501).
 */
class WordwolfJudgeLogicTests {

    private val logic = WordwolfJudgeLogic()

    private fun summaries(
        state: SimulationState,
        language: String,
    ): List<String> {
        val events = mutableListOf<SimulationEvent>()
        logic.calculate(state, language) { events += it }
        return events.filterIsInstance<SimulationEvent.Summary>().map { it.text }
    }

    @Test
    fun majorityWinsWhenWolfMostVoted() {
        val state = SimulationState(
            voteResults = mapOf("Wolf" to 3, "Other" to 1),
            variables = mapOf("wolf_name" to "Wolf"),
        )
        val out = summaries(state, "ja")
        assertEquals(1, out.size)
        // Rendered-string parity (ja, wolf found).
        assertEquals("最多得票: Wolf (3票) — 多数派の勝ち！ウルフを見破った！", out[0])
    }

    @Test
    fun wolfWinsWhenNotDetected() {
        val state = SimulationState(
            voteResults = mapOf("Innocent" to 3, "Wolf" to 1),
            variables = mapOf("wolf_name" to "Wolf"),
        )
        assertTrue(summaries(state, "ja")[0].contains("ウルフの勝ち"))
    }

    @Test
    fun tieResolvesToStableWinner() {
        // Canonical tie-break (count desc, name desc) → Bob wins; Bob != wolf
        // (Alice) → wolf escapes deterministically on every call.
        fun run(): String {
            val state = SimulationState(
                voteResults = mapOf("Alice" to 2, "Bob" to 2),
                variables = mapOf("wolf_name" to "Alice"),
            )
            return summaries(state, "ja").firstOrNull() ?: ""
        }
        val first = run()
        assertTrue(first.contains("Bob"))
        assertTrue(first.contains("ウルフの勝ち"))
        assertFalse(first.contains("Alice"))
        assertEquals(first, run())
        assertEquals(first, run())
    }

    @Test
    fun handlesEmptyVoteResults() {
        // Both empty-votes literals pinned (rendered-string parity), symmetric
        // with the win/lose templates below.
        assertEquals("投票結果がありません", summaries(SimulationState(), "ja")[0])
        assertEquals("No votes recorded", summaries(SimulationState(), "en")[0])
    }

    @Test
    fun majorityWinsInEnglish() {
        val state = SimulationState(
            voteResults = mapOf("Wolf" to 3, "Other" to 1),
            variables = mapOf("wolf_name" to "Wolf"),
        )
        val out = summaries(state, "en")
        assertEquals(1, out.size)
        // Rendered-string parity (en, wolf found).
        assertEquals("Most votes: Wolf (3) — Majority wins! The wolf was found.", out[0])
        assertFalse(out[0].contains("多数派の勝ち"))
    }

    @Test
    fun wolfWinsInEnglish() {
        val state = SimulationState(
            voteResults = mapOf("Innocent" to 3, "Wolf" to 1),
            variables = mapOf("wolf_name" to "Wolf"),
        )
        val out = summaries(state, "en")
        assertEquals(1, out.size)
        assertEquals("Most votes: Innocent (3) — The wolf wins! Escaped detection.", out[0])
        assertFalse(out[0].contains("ウルフの勝ち"))
    }
}
